<a href="https://colab.research.google.com/github/AldarXV/DA_projects/blob/main/case20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install dataprep

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
from dataprep.eda import create_report
import numpy as np
import pandas as pd

NumExpr defaulting to 2 threads.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
df=pd.read_csv('/content/drive/MyDrive/кейс_20.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377741 entries, 0 to 377740
Data columns (total 20 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   ProductName           377741 non-null  object 
 1   ProductCategory       377741 non-null  object 
 2   ItemGroup             377741 non-null  object 
 3   KitType               377741 non-null  object 
 4   Channels              377741 non-null  int64  
 5   Demographic           377741 non-null  object 
 6   RetailPrice           377741 non-null  float64
 7   OrderNumber           377741 non-null  object 
 8   OrderDate             377741 non-null  object 
 9   Quantity              377741 non-null  int64  
 10  UnitPrice             377741 non-null  float64
 11  DiscountAmount        377741 non-null  float64
 12  SaleAmount            377741 non-null  float64
 13  OrderProcessing       377741 non-null  int64  
 14  StateName             377741 non-null  object 
 15  

В таблице 20 столбцов и 377741 строк. Пропущенных данных нет. Изменим тип данных для OrderDate

In [ ]:
df['OrderDate'] = pd.to_datetime(df['OrderDate'], infer_datetime_format=True)

#EDA и подготовка данных

In [ ]:
#create_report(df).show()

1. Следует удалить столбцы DiscountAmount, OrderNumber, так как они неинформативны.
2. RetailPrice, UnitPrice, SaleAmount - переменные коллинеарны, следует удалить столбцы RetailPrice, UnitPrice.

In [ ]:
df=df.drop(['DiscountAmount', 'OrderNumber', 'RetailPrice', 'UnitPrice', 'ABC', 'XYZ'], axis=1)

#Регрессия

In [ ]:
df['Year']=df['OrderDate'].dt.year
df['Month']=df['OrderDate'].dt.month

In [ ]:
df.head()

ProductName ProductCategory ItemGroup KitType  Channels  \
0  Piper Cub 3 Channel         Trainer  Airplane     RTF         3   
1  Piper Cub 3 Channel         Trainer  Airplane     RTF         3   
2  Piper Cub 3 Channel         Trainer  Airplane     RTF         3   
3  Piper Cub 3 Channel         Trainer  Airplane     RTF         3   
4  Piper Cub 3 Channel         Trainer  Airplane     RTF         3   

  Demographic  OrderDate  Quantity  SaleAmount  OrderProcessing   StateName  \
0    Beginner 2017-01-01         1       69.95                8     Indiana   
1    Beginner 2017-01-02         1       69.95                2    Illinois   
2    Beginner 2017-01-06         1       69.95                3        Ohio   
3    Beginner 2017-01-08         1       69.95                3     Arizona   
4    Beginner 2017-01-08         1       69.95                2  California   

  RegionName ABC-XYZ OrderProcessingSpeed  Year  Month  
0    Midwest      CX              Expired  2017      1  
1    Midwest      CX               OnTime  2017      1  
2    Midwest      CX              Expired  2017      1  
3   Southern      CX              Expired  2017      1  
4  Southwest      CX               OnTime  2017      1

In [ ]:
df=df.drop('OrderDate', axis=1)

In [ ]:
df

ProductName   ProductCategory   ItemGroup  \
0                        Piper Cub 3 Channel           Trainer    Airplane   
1                        Piper Cub 3 Channel           Trainer    Airplane   
2                        Piper Cub 3 Channel           Trainer    Airplane   
3                        Piper Cub 3 Channel           Trainer    Airplane   
4                        Piper Cub 3 Channel           Trainer    Airplane   
...                                      ...               ...         ...   
377736                     3CAX-B Helicopter          Co-Axial  Helicopter   
377737                   Piper Cub 4 Channel           Trainer    Airplane   
377738  Tailspin Heli - Max Pro Flight - 6ch  Collective pitch  Helicopter   
377739  Tailspin Heli - Max Pro Flight - 6ch  Collective pitch  Helicopter   
377740                     3CAX-B Helicopter          Co-Axial  Helicopter   

       KitType  Channels   Demographic  Quantity  SaleAmount  OrderProcessing  \
0          RTF         3      Beginner         1       69.95                8   
1          RTF         3      Beginner         1       69.95                2   
2          RTF         3      Beginner         1       69.95                3   
3          RTF         3      Beginner         1       69.95                3   
4          RTF         3      Beginner         1       69.95                2   
...        ...       ...           ...       ...         ...              ...   
377736     RTF         3        Novice         1       63.55                0   
377737     RTF         4  Intermediate         1      193.55                0   
377738     KIT         6  Professional         1      560.25                0   
377739     KIT         6  Professional         1      579.55                0   
377740     RTF         3        Novice         1       63.55                0   

           StateName         RegionName ABC-XYZ OrderProcessingSpeed  Year  \
0            Indiana            Midwest      CX              Expired  2017   
1           Illinois            Midwest      CX               OnTime  2017   
2               Ohio            Midwest      CX              Expired  2017   
3            Arizona           Southern      CX              Expired  2017   
4         California          Southwest      CX               OnTime  2017   
...              ...                ...     ...                  ...   ...   
377736  South Dakota            Midwest      BY               OnTime  2019   
377737      Oklahoma            Midwest      AX               OnTime  2019   
377738       Georgia           Southern      AY               OnTime  2019   
377739    Washington  Pacific Northwest      AY               OnTime  2019   
377740     Wisconsin            Midwest      BY               OnTime  2019   

        Month  
0           1  
1           1  
2           1  
3           1  
4           1  
...       ...  
377736     12  
377737     12  
377738     12  
377739     12  
377740     12  

[377741 rows x 15 columns]

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
df1 = df[['ProductName','ProductCategory','ItemGroup','KitType','Channels','Demographic', 'RegionName', 'StateName','SaleAmount', 'Year',	'Month']]
df1=df1.groupby(['ProductName','ProductCategory','ItemGroup', 'KitType','Channels','Demographic', 'RegionName', 'StateName', 'Year','Month'], as_index=False ).mean()


In [ ]:
df1_dummies = pd.get_dummies(df1)
df1_dummies

Channels  Year  Month  SaleAmount  ProductName_3CAX-B Helicopter  \
0             3  2017      3       49.95                              1   
1             3  2017      4       49.95                              1   
2             3  2017      5       49.95                              1   
3             3  2017      6       39.95                              1   
4             3  2017      7       49.95                              1   
...         ...   ...    ...         ...                            ...   
27813         4  2019      3      196.95                              0   
27814         4  2019      5      196.95                              0   
27815         4  2019      6      196.95                              0   
27816         4  2019     11      216.65                              0   
27817         4  2019     12      216.65                              0   

       ProductName_3CFP-I Helicopter  ProductName_4CAX-B Helicopter  \
0                                  0                              0   
1                                  0                              0   
2                                  0                              0   
3                                  0                              0   
4                                  0                              0   
...                              ...                            ...   
27813                              0                              0   
27814                              0                              0   
27815                              0                              0   
27816                              0                              0   
27817                              0                              0   

       ProductName_4CFP-I Helicopter  ProductName_6CCP-A Helicopter  \
0                                  0                              0   
1                                  0                              0   
2                                  0                              0   
3                                  0                              0   
4                                  0                              0   
...                              ...                            ...   
27813                              0                              0   
27814                              0                              0   
27815                              0                              0   
27816                              0                              0   
27817                              0                              0   

       ProductName_P47 4 Channel  ...  StateName_South Dakota  \
0                              0  ...                       0   
1                              0  ...                       0   
2                              0  ...                       0   
3                              0  ...                       0   
4                              0  ...                       0   
...                          ...  ...                     ...   
27813                          0  ...                       0   
27814                          0  ...                       0   
27815                          0  ...                       0   
27816                          0  ...                       0   
27817                          0  ...                       0   

       StateName_Tennessee  StateName_Texas  StateName_Utah  \
0                        0                0               0   
1                        0                0               0   
2                        0                0               0   
3                        0                0               0   
4                        0                0               0   
...                    ...              ...             ...   
27813                    0                0               0   
27814                    0                0               0   
27815                    0                0         

In [ ]:
df1_dummies=df1_dummies.sort_values(by=['Year', 'Month'])
df1_dummies

Channels  Year  Month  SaleAmount  ProductName_3CAX-B Helicopter  \
10058         3  2017      1       69.95                              0   
10094         3  2017      1       69.95                              0   
10409         3  2017      1       69.95                              0   
10619         3  2017      1       69.95                              0   
10791         3  2017      1       69.95                              0   
...         ...   ...    ...         ...                            ...   
27713         4  2019     12      216.65                              0   
27740         4  2019     12      216.65                              0   
27762         4  2019     12      208.75                              0   
27795         4  2019     12      216.65                              0   
27817         4  2019     12      216.65                              0   

       ProductName_3CFP-I Helicopter  ProductName_4CAX-B Helicopter  \
10058                              0                              0   
10094                              0                              0   
10409                              0                              0   
10619                              0                              0   
10791                              0                              0   
...                              ...                            ...   
27713                              0                              0   
27740                              0                              0   
27762                              0                              0   
27795                              0                              0   
27817                              0                              0   

       ProductName_4CFP-I Helicopter  ProductName_6CCP-A Helicopter  \
10058                              0                              0   
10094                              0                              0   
10409                              0                              0   
10619                              0                              0   
10791                              0                              0   
...                              ...                            ...   
27713                              0                              0   
27740                              0                              0   
27762                              0                              0   
27795                              0                              0   
27817                              0                              0   

       ProductName_P47 4 Channel  ...  StateName_South Dakota  \
10058                          0  ...                       0   
10094                          0  ...                       0   
10409                          0  ...                       0   
10619                          0  ...                       0   
10791                          0  ...                       0   
...                          ...  ...                     ...   
27713                          0  ...                       0   
27740                          0  ...                       0   
27762                          0  ...                       0   
27795                          0  ...                       0   
27817                          0  ...                       0   

       StateName_Tennessee  StateName_Texas  StateName_Utah  \
10058                    0                0               0   
10094                    0                0               0   
10409                    0                0               0   
10619                    0                0               0   
10791                    0                0               0   
...                    ...              ...             ...   
27713                    0                1               0   
27740                    0                0               0   
27762                    0                0         

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df1_dummies.drop(columns='SaleAmount').to_numpy(), df1_dummies['SaleAmount'].to_numpy(), test_size = 0.3, random_state = 1)

In [ ]:
gradient_boosted = GradientBoostingRegressor(random_state=2)
gradient_boosted.fit(X_train, y_train)
model_pred = gradient_boosted.predict(X_test)
print('mae= ', np.mean(abs(model_pred- y_test)))

mae=  4.9529227239756235


Значение mae=4.953 больше полученного в Knime (2.899). 

In [ ]:
np.max(df1_dummies['SaleAmount'])

579.55

#Классификация

In [ ]:
df.columns

Index(['ProductName', 'ProductCategory', 'ItemGroup', 'KitType', 'Channels',
       'Demographic', 'Quantity', 'SaleAmount', 'OrderProcessing', 'StateName',
       'RegionName', 'ABC-XYZ', 'OrderProcessingSpeed', 'Year', 'Month'],
      dtype='object')

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [ ]:
df2 = df[['ProductName','ProductCategory','ItemGroup','KitType','Channels','Demographic', 'RegionName', 'StateName', 'ABC-XYZ', 'OrderProcessingSpeed']]

In [ ]:
df2_dummies = pd.get_dummies(df2)

In [ ]:
df2_dummies.columns

Index(['Channels', 'ProductName_3CAX-B Helicopter',
       'ProductName_3CFP-I Helicopter', 'ProductName_4CAX-B Helicopter',
       'ProductName_4CFP-I Helicopter', 'ProductName_6CCP-A Helicopter',
       'ProductName_P47 4 Channel', 'ProductName_P47 5 Channel',
       'ProductName_P51', 'ProductName_Piper Cub 3 Channel',
       'ProductName_Piper Cub 4 Channel', 'ProductName_SkyTrainer',
       'ProductName_Tailspin Aviator Mk2-11',
       'ProductName_Tailspin Aviator Mk2-12',
       'ProductName_Tailspin Aviator Mk2-15',
       'ProductName_Tailspin Heli - Co-Ax Pro Mk I - 4ch',
       'ProductName_Tailspin Heli - Max Pro Flight - 6ch',
       'ProductName_Tailspin Heli - Pro Mk III - 5ch',
       'ProductName_Tailspin Warbird BM32',
       'ProductName_Trainer - Tailspin GL-120',
       'ProductName_Trainer - Tailspin GL-155', 'ProductCategory_Co-Axial',
       'ProductCategory_Collective pitch', 'ProductCategory_Fixed pitch',
       'ProductCategory_Glider', 'ProductCategory_Train

In [ ]:
df2_dummies=df2_dummies.drop('OrderProcessingSpeed_Expired', axis=1)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df2_dummies.drop(columns='OrderProcessingSpeed_OnTime').to_numpy(), df2_dummies['OrderProcessingSpeed_OnTime'].to_numpy(), test_size = 0.3, random_state = 1)

In [ ]:
logReg = LogisticRegression(max_iter=300)
logReg.fit(X_train, y_train)
print('Точность')
print('Логистическая регрессия: ', round(logReg.score(X_test, y_test), 4))

random_forest = RandomForestClassifier(random_state=1)
random_forest.fit(X_train, y_train)
print('Random forest', round(random_forest.score(X_test, y_test), 4))

Точность
Логистическая регрессия:  0.5612
Random forest 0.5568
